### ISSUE TITLE
Game Keyboard Shortcuts

### ISSUE THREAD

**Issue by krs0** [vcmi/vcmi#1834]
### Global Game shortcuts

- [ ] `ENTER` - Confirmation
- [ ] `ESCAPE` - Cancel
- [ ] `E` - End turn
- [ ] `L` - Load game
- [ ] `S` - Save game
- [ ] `CTRL + R` - restart game with Yes/No window
- [ ] `F5` - Toolbar?
- [ ] `Hold CTRL` while in window mode - disable map scrolling when you leave the window

### Adventure screen shortcuts

- [ ] `SPACE` - move hero to selected tile / revisit current location
- [ ] `Cursor keys` - move hero on adventure map
- [ ] `Hold ALT + hover mouse` - See movement points required to reach map tile
- [ ] `SHIFT + Left Click` on hero - Make clicked hero the active one (Instead of default hero meetup)
- [ ] `ENTER` - open the selected hero's info window / open the selected town's window
- [ ] `T` - open first listed town's window
- [ ] `M` - open first Market Place from a town
- [ ] `G` - open first Thieves Guild from a town with tavern
- [ ] `I` - Scenario information
- [ ] `P` - Puzzle map
- [ ] `V` - View World
- [ ] `D` - Dig
- [ ] `C` - Adventure map casting
- [ ] `ESCAPE` - cancel Adventure map casting

### Heros / Towns Panel

- [ ] `ALT + Mouse Drag` - rearrange heroes/towns

### Combat screen

- [ ] `W` - Wait
- [ ] `D` - Defend
- [ ] `G` - Ground Area attack (with Gogs and Liches), Fairy Dragons
- [ ] `C` - open Spell book
- [ ] `I` - Show hero stats permanently
- [ ] `CTRL + hover mouse` unit - See unit's general movement
- [ ] `SHIFT + hover mouse` unit - See exact unit's movement (calculate obstacles)
- [ ] `Q` - Finish combat with Quick Combat

### Town Screen

- [ ] `SPACE` - Move hero from garrison into town, and vice versa
- [ ] `F10` - Switch armies from garrison into town, and vice versa
- [ ] `B` - Open build menu (town hall)
- [ ] `R` - Open recruit creatures menu (fort click)
- [ ] `M` - Open marketplace (B in HotA)
- [ ] `CTRL + Click` creature portrait in town screen - Immediately buy all creatures of that type (available in town)
- [ ] `CTRL + Click` town icon (to the left of income icon) to instantly buy ALL available creatures in town

##### On Stack

- [ ] `ALT + SHIFT + Click` - dismiss stack
- [ ] `CTRL + Click `- splits current stack to next empty slot, if possible
- [ ] `CTRL + SHIFT + Click` - splits from current stack, stacks of 1 in all free slots
- [ ] `ALT + Click` - merges all units of same type into 1 stack
- [ ] `SHIFT + Click` - progressively splits the current stack into equal stacks on available slots (empty and occupied by same unit), but not over 1 unit slots of that type. (you can do multiple clicks to divide the stack further)

##### On Artifact

- [ ] `ALT + Click` - move the artifact from the doll to the backpack and from the backpack to the doll
- [ ] `CTRL + Click` - transfer the artifact to the other hero and wear it
- [ ] `SHIFT + Click` - transfer the artifact to the backpack of the other hero

### Town screen / Hero meetup screen

- [ ] `F10` - full army exchange between heroes
- [ ] `ALT + CTRL + Click` - transfers whole stack to other hero, filling the left most slot, or putting it over an existing stack of same type
- [ ] `ALT + CTRL + SHIFT + Click` - transfers full army except 1 unit of current selected stack

### Hero meetup screen

- [ ] `CTRL + Click` on transfer artifacts button - transfer doll artifacts only
- [ ] `SHIFT + Click` on transfer artifacts button - transfer backpack artifacts only

**Comment by kambala-decapitator** [comment]
related: #1382

**Comment by krs0** [comment]
@kambala-decapitator  Combine and discard that one?

**Comment by kambala-decapitator** [comment]
> @kambala-decapitator Combine and discard that one?

I'm fine with that

**Comment by krs0** [comment]
From [kambala-decapitator](https://github.com/kambala-decapitator) for debate :)
"purchase buildings" screen can also receive hotkeys, e.g. 1-7 keys to build/upgrade respective dwelling

**Comment by IvanSavenko** [comment]
> Space - move hero to selected tile (Currently M key is used)
H - open first listed hero's window (Currently Next Hero)

Not sure if we should change shortcuts that were in H3

> G - Ground Area attack / Cast (with Gogs and Liches), Fairy Dragons

I think we already have that but at 'F'

> Left, Right - selects left/right choice in dialogs (new level, arena, etc.)

I believe we have 1/2/3... for these. IMO might be better since works for 3+ components.

**Comment by krs0** [comment]
Regarding changes to original H3, or even HotA shortcuts... 99% of the time they should stay as they are so players do not get surprises, but!
Game evolved a lot in functionality, and some things can be corrected.
AFAIK 2 keys (99% out of 150 keys :) ) were proposed to be changed: M, and H for Next Hero.

Regarding M key for move hero... Space key fills that functionality already with revisit object. M is natural for Marketplace, on adventure map and in Town screen.
By consolidating hero move to Space, M key is freed up for Marketplace, which in turn frees B (HotA Marketplace) for morè natural Build (in Town menu).

M opens Marketplace, G opens Guild, T opens first town, H opens what...? first hero :) Alt H cycles heroes, Alt T cycles towns.
Opening first hero will be more useful once heroes can be rearanged. You will have a one Key shortcut to go to view your main.

**Comment by IvanSavenko** [comment]
If we're OK with rearranging shortcuts a bit then:

1) Market & Space behavior:
- `W` -> `Z` - Wake Hero - merged with Sleep, making it into a toggle
- `V` -> `W` - View World
- `V` - Visit Current Object
- `Space` - Move Hero
- `M` - Open Market

2) Lists scrolling:
`H` - Select First Hero
`T` - Select First Town
`N` - Select Next (Town or Hero, context-depending)
`Enter` - Open Selected

Also, I think we should minimize shortcuts with modifiers keys - Alt, Ctrl, Shift & use them only for "advanced" functionality - something that casual players are unlikely to use.

**Comment by krs0** [comment]
I like the proposals... its simple and clean! will update issue.
I would keep Space as secondary option for revisit. It is likenthat since forever and it does not disturb your proposal.

**Comment by IvanSavenko** [comment]
In that case, here is how adventure map keyboard layout looks like:

(your list is nice, but hard to visualize & find unused buttons)
![Shortcuts](https://user-images.githubusercontent.com/1576820/230716575-3122329e-1054-47d5-8ac0-dd9de1731f1c.png)

**Comment by krs0** [comment]
Updated list with your suggestions, also what you said about Ctrl,Alt makes sense.... I would still keep them available, but will group those at some point as Advanced Sortcuts.

**Comment by Laserlicht** [comment]
Maybe we can adapt some things from [SOD_SP](https://docs.google.com/document/d/1JlQ6TC97d_Bb1g_sDRpxTvkKHtyXgZ3qORG5LJS8tp8/edit)?

For example quick save/load, quick spell?

**Comment by krs0** [comment]
Quickspell is a nice one.
You save some 9 spells, they appear always besides your combat screen, and you can cast them by pressing the according number key

**Comment by luk3ZPL** [comment]
+ Backspace reset zoom.

**Comment by krs0** [comment]
Added ty.

**Comment by swats-the-floran** [comment]
Hero meetup screen - may i suggest closing the screen by Esc?

**Comment by IvanSavenko** [comment]
That's not possible at the moment?
Makes sense. IMO any screen should be possible to close via both Enter and Escape

**Comment by swats-the-floran** [comment]
> That's not possible at the moment? Makes sense. IMO any screen should be possible to close via both Enter and Escape

totally agree. i will add other screens here if i find any.

**Comment by swats-the-floran** [comment]
Screens that user can't close by Esc atm:
- View World
- Quest Log
- Hero Exchange

**Comment by Laserlicht** [comment]
Would also great if such windows could closed by clicking outside window (better experience on touch screen).

**Comment by kambala-decapitator** [comment]
one more suggestion: shortcuts to change unit selection for purchase

<img width="473" alt="Screenshot 2024-08-04 at 16 08 39" src="https://github.com/user-attachments/assets/20207889-a7a7-4b9c-a96f-1e15c1e89f92">

**Comment by AlexSnowLeo** [comment]
Hi there!
I installed VCMI on my Steam Deck not too long ago and am trying to adjust the gamepad layout for good game control. The idea arose to assign hotkeys to the buttons of the gamepad. And it wasn't bad to have information about them right in the game. For example, in pop-up windows on these buttons - it would be ideal. What do you think?
PS: This can be done not only in the game itself, but also in the main menu.

**Comment by Laserlicht** [comment]
> Hi there! I installed VCMI on my Steam Deck not too long ago and am trying to adjust the gamepad layout for good game control. The idea arose to assign hotkeys to the buttons of the gamepad. And it wasn't bad to have information about them right in the game. For example, in pop-up windows on these buttons - it would be ideal. What do you think? PS: This can be done not only in the game itself, but also in the main menu.

There is an Keybinding dialog in Develop version: #5937

Also supports gamepad bindings.

**Comment by AlexSnowLeo** [comment]
> There is an Keybinding dialog in Develop version: [#5937](https://github.com/vcmi/vcmi/pull/5937)
> 
> Also supports gamepad bindings.

Yes, I've seen this setup dialog - it's a great solution. But I mean, here's an add-on, to show the Hot Key in a pop-up window, if available:

<img width="472" height="523" alt="Image" src="https://github.com/user-attachments/assets/bac6e174-7b86-472c-9bd4-e03ae221ba4e" />
